### PR TITLE
Group parsing

### PR DESCRIPTION
**Pull request by InnocentZero**
Adds TOML parsing. Some backend traits need changes to make them TOML compatible. Else we'd need to change the group file description language.

### CODE REVIEW

**Comment by InnocentZero** [comment]
In particular, unit type `()` doesn't work. It has been used in a lot of places so we either need to find alternatives for backend types or find a different config file format.

File: src/groups.rs [hunk 1]
```diff
@@ -22,17 +26,37 @@ impl Groups {
...
+                .ok_or(anyhow!("Will not fail on Linux"))?
+                .to_string();
+            println!("group_name: {group_name}");
+            let file_contents = read_to_string(group_file).context("Reading group file")?;
+
+            let packages: PackagesInstall = toml::from_str(&file_contents)?;
+            groups.insert(group_name, packages);
+        }
+        Ok(groups)
     }
```
Comment (ripytide) [inline src/groups.rs]:
```suggestion
    /// Loads Groups from a users pacdef config folder.
    pub fn load(group_dir: &Path) -> Result<Self> {
        let mut groups = Self::default();

        let group_dir = group_dir.join("groups/");
        if !group_dir.is_dir() {
            return Err(anyhow!(
                "the groups directory was not found in the pacdef config folder, please create it"
            ));
        }

        let group_files: Vec<DirEntry> = WalkDir::new(&group_dir)
            .follow_links(true)
            .into_iter()
            .collect::<Result<_, _>>()?;

        for group_file in group_files.into_iter().filter(|x| x.path().is_file()) {
            let group_name = group_file
                .path()
                .strip_prefix(&group_dir)?
                .to_str()
                .context("converting group file name to a string")?
                .to_string();

            log::info!("parsing group file: {group_name}@{group_file:?}");

            let file_contents = read_to_string(group_file.path()).context("reading group file")?;

            let packages: PackagesInstall = toml::from_str(&file_contents)?;
            groups.insert(group_name, packages);
        }

        Ok(groups)
    }
```

It's probably best we get the user to change their config directory by creating a groups folder than attempting to modify it ourselves.

**Comment by ripytide** [comment]
I think the best option so we don't have to throw away the strongly typed backend associated trait types might be to write a bit of custom serialization/deserialization logic so that something like this gets mapped into the `BTreeMaps` inside `PackagesInstall`:
```toml
arch = [
"pacdef-bin",
{ package="neovim", clean_install = true },
]
```

**Comment by ripytide** [comment]
Semantically this is roughly equivalent to a `BTreeSet<struct { package: String, clean_install: Option<bool>,}` but then we lose the uniqueness guarantee from the `BTreeMap`

**Comment by ripytide** [comment]
I rebased the PR after #3 was merged

**Comment by ripytide** [comment]
I think we might need to finalize our ideal file format first in #5 and then implementing it will probably be best via a custom parsing function that builds a struct from a generic `toml::Table` type without serde due to the difficulties with getting serde to serialize/deserialize exactly the way we want.

**Comment by ripytide** [comment]
The PR looks good, I'm going to merge this so I can take a shot at implementing the file format from #5 in another PR

**Pull request merged**